File: GameClient/src/backend/parser.py
from typing import List

from src.shared.constants import CMessagePartsSizes, CMessageConfig, CCommandTypeEnum, Param, NetworkMessage, \
    MessageFormatError


def parse_message(input_str: str) -> NetworkMessage:
    def _parse_param_str(params_string)->Param:
        def __process_array_values(param : Param) -> Param:
            def ___is_array_param(param : Param) -> bool:
                if len(param.value) == 0:
                    return False
                return param.value[0] == CMessageConfig.ARRAY_BRACKETS.opening and param.value[-1] == CMessageConfig.ARRAY_BRACKETS.closing
            def ___parse_array_element(array_element: str)-> List[Param]:

                param_list = []
                # {"gameName": "Game1", "maxPlayers": "4", "connectedPlayers": "2"}
                if array_element[0] != CMessageConfig.PARAMS_BRACKETS.opening or array_element[-1] != CMessageConfig.PARAMS_BRACKETS.closing:
                    raise ValueError("Invalid array element format")
                array_element = array_element[1:-1]
                # "gameName": "Game1", "maxPlayers": "4", "connectedPlayers": "2"
                values_array = array_element.split(CMessageConfig.PARAMS_DELIMITER)
                for value in values_array:
                    param_list.append(__parse_param_element(value))

                return param_list
            # if not array
            if not ___is_array_param(param):
                return param

            array_str = param.value[1:-1]
            if len(array_str) == 0:
                return Param(param.name, [])

            array = array_str.split(CMessageConfig.PARAMS_ARRAY_DELIMITER)

            value_array = []
            for element in array:
                value_array.append(___parse_array_element(element))

            return Param(param.name,value_array)
        def __parse_param_element(str):
            if len(str) == 0:
                return Param("", "")

            param_key_value = str.split(CMessageConfig.PARAMS_KEY_VALUE_DELIMITER, 1)

            name = param_key_value[0]
            if name[0] != CMessageConfig.PARAMS_WRAPPER or name[-1] != CMessageConfig.PARAMS_WRAPPER:
                raise ValueError("Invalid param format")
            name = name[1:-1]

            value = param_key_value[1]
            if value[0] != CMessageConfig.PARAMS_WRAPPER or value[-1] != CMessageConfig.PARAMS_WRAPPER:
                raise ValueError("Invalid param format")
            value = value[1:-1]

            return Param(name, value)

        if len(params_string) == 0:
            return Param("", "")

        if params_string[0] != CMessageConfig.PARAMS_BRACKETS.opening:
            raise ValueError("Invalid paramArray format")

        if params_string[-1] != CMessageConfig.PARAMS_BRACKETS.closing:
            raise ValueError("Invalid paramArray format")

        # without brackets
        params_string = params_string[1:-1] #always one parameter

        parameter = __parse_param_element(params_string)

        param = __process_array_values(parameter)

        return param

    def _convert_arrayparam_to_specified_datastructures(command_id: int, param: Param) -> List[Param]:
        command = CCommandTypeEnum.get_command_by_id(command_id)
        message_param_list_info = command.param_list_info
        if message_param_list_info is None:
            return [param]

        param_value = param.value

        if len(param_value) == 0:
            return [param]
        if not isinstance(param_value, list):
            raise [param]

        param_array = param_value
        # check names in array elements
        for element in param_array:
            elements_names = [param.name for param in element]
            if elements_names != message_param_list_info.param_names:
                raise MessageFormatError("Invalid parameter names in array element")
        converted_data = message_param_list_info.convert_function(param_array)

        return [Param(param.name, converted_data)]


    def _parse_message_player_nickname(input):
        player_nickname = ""
        player_nickname_size = 0

        if len(input) == 0:
            raise ValueError("Invalid player ID format")

        opening = CMessageConfig.PARAMS_BRACKETS.opening
        closing = CMessageConfig.PARAMS_BRACKETS.closing

        player_nickname = input[len(opening):input.find(closing)]
        player_nickname_size = len(player_nickname) + len(opening) + len(closing)

        return player_nickname, player_nickname_size

    if len(input_str) == 0:
        raise ValueError("Invalid input format")

    signature_size = CMessagePartsSizes.SIGNATURE_SIZE
    command_id_size = CMessagePartsSizes.COMMANDID_SIZE
    timestamp_size = CMessagePartsSizes.TIMESTAMP_SIZE
    min_size_message = signature_size + command_id_size + timestamp_size

    if len(input_str) < min_size_message:
        raise ValueError("Invalid input format")

    start = 0

    # Check if not more end of message characters
    if input_str[:-1].find(CMessageConfig.END_OF_MESSAGE) != -1:
        raise ValueError("Buffer read multiple messages")

    # Read Signature
    signature = input_str[start : start + signature_size]
    if signature != CMessageConfig.SIGNATURE:
        raise ValueError("Invalid signature")
    start += signature_size

    # Read Command ID
    command_id = input_str[start : start + command_id_size]
    if not command_id.isdigit() or int(command_id) < 0 or not CCommandTypeEnum.is_command_id_in_enum(int(command_id)):
        raise ValueError("Invalid command ID")
    start += command_id_size

    # Read Timestamp
    timestamp = input_str[start : start + timestamp_size]
    start += timestamp_size

    # Read nickname
    player_nickname, player_nickname_size = _parse_message_player_nickname(input_str[start:])
    start += player_nickname_size

    # Read Parameters
    parameters_str = input_str[start:]

    param = _parse_param_str(parameters_str)

    # Convert values
    command_id_int = int(command_id)

    param = _convert_arrayparam_to_specified_datastructures(command_id_int, param)

    return NetworkMessage(signature, command_id_int, timestamp, player_nickname, param)

def convert_message_to_network_string(message: NetworkMessage) -> str:
    def _convert_params_to_network_string(params: List[Param]) -> str:
        params_str = CMessageConfig.PARAMS_BRACKETS.opening
        for param in params:
            param_name = str(param.name)
            param_value = str(param.value)
            params_str += CMessageConfig.PARAMS_WRAPPER + param_name + CMessageConfig.PARAMS_WRAPPER + CMessageConfig.PARAMS_KEY_VALUE_DELIMITER + CMessageConfig.PARAMS_WRAPPER + param_value + CMessageConfig.PARAMS_WRAPPER + CMessageConfig.PARAMS_DELIMITER
        params_str = params_str.rstrip(CMessageConfig.PARAMS_DELIMITER)  # remove the trailing delimiter
        params_str += CMessageConfig.PARAMS_BRACKETS.closing

        return params_str

    network_string = ""

    signature_size = CMessagePartsSizes.SIGNATURE_SIZE
    command_id_size = CMessagePartsSizes.COMMANDID_SIZE
    timestamp_size = CMessagePartsSizes.TIMESTAMP_SIZE

    # Signature
    if len(message.signature) != signature_size:
        raise ValueError("Invalid signature")
    network_string += message.signature

    # Command ID
    command_id_str = str(format(message.command_id, '02d'))
    if len(command_id_str) != command_id_size:
        raise ValueError("Invalid command ID")
    network_string += command_id_str

    # Timestamp
    if len(message.timestamp) != timestamp_size:
        raise ValueError("Invalid timestamp")
    network_string += message.timestamp

    # PlayerNickname
    network_string += CMessageConfig.PARAMS_BRACKETS.opening
    network_string += message.player_nickname
    network_string += CMessageConfig.PARAMS_BRACKETS.closing

    # Parameters
    params_str = _convert_params_to_network_string(message.parameters)
    network_string += params_str

    # Add end of message
    network_string += CMessageConfig.END_OF_MESSAGE

    return network_string


def convert_list_cube_values_to_network_string(selected_cubes: List[int]) -> str:
    # [{"gameName":"Game1","maxPlayers":"4","connectedPlayers":"2"};{"gameName":"Game2","maxPlayers":"4","connectedPlayers":"2"}]
    name_value = "value"

    selected_cubes_str = CMessageConfig.ARRAY_BRACKETS.opening
    for cube in selected_cubes:
        selected_cubes_str += CMessageConfig.PARAMS_BRACKETS.opening
        selected_cubes_str += CMessageConfig.PARAMS_WRAPPER + name_value + CMessageConfig.PARAMS_WRAPPER
        selected_cubes_str += CMessageConfig.PARAMS_KEY_VALUE_DELIMITER
        selected_cubes_str += CMessageConfig.PARAMS_WRAPPER + str(cube) + CMessageConfig.PARAMS_WRAPPER
        selected_cubes_str += CMessageConfig.PARAMS_BRACKETS.closing
        selected_cubes_str += CMessageConfig.PARAMS_ARRAY_DELIMITER
    selected_cubes_str = selected_cubes_str.rstrip(
        CMessageConfig.PARAMS_ARRAY_DELIMITER)  # remove the trailing delimiter
    selected_cubes_str += CMessageConfig.ARRAY_BRACKETS.closing

    return selected_cubes_str

File: GameServer/internal/parser/parser.go
package parser

import (
	"fmt"
	"gameserver/internal/logger"
	"gameserver/internal/models"
	"gameserver/internal/utils/constants"
	"gameserver/internal/utils/errorHandeling"
	"reflect"
	"strconv"
	"strings"
	"unsafe"
) //todo predelat bez pouziti regex

//region FUNCTIONS PARSE

func ParseMessage(input string) (models.Message, error) {
	logger.Log.Debug("Parsing message: %v", input)

	parts := strings.Split(input, constants.CMessageEndDelimiter)
	// if len is not 2 and second part is not empty
	if len(parts) != 2 && len(parts[1]) != 0 {
		err := fmt.Errorf("invalid input format")
		errorHandeling.PrintError(err)
		return models.Message{}, err
	}

	input = parts[0]

	if len(input) == 0 {
		err := fmt.Errorf("empty input")
		errorHandeling.PrintError(err)
		return models.Message{}, err
	}

	messageData := models.MessageHeader{}
	signatureSize := int(unsafe.Sizeof(messageData.Signature))
	commandIDSize := int(unsafe.Sizeof(messageData.CommandID))
	timeStempSize := int(unsafe.Sizeof(messageData.TimeStamp))
	minSizeMessage := signatureSize + commandIDSize + timeStempSize

	if len(input) < minSizeMessage {
		return models.Message{}, fmt.Errorf("invalid input format")
	}

	start := 0

	//Read Signature
	signature := input[start : start+signatureSize]
	start += signatureSize

	//Read Command ID
	commandID := input[start : start+commandIDSize]
	start += commandIDSize

	//Read TimeStemp
	timeStamp := input[start : start+timeStempSize]
	start += timeStempSize

	//Read nickname
	playerNickname, playerNicknameSize, err := parseMessagePlayerID(input[start:])
	if err != nil {
		errorHandeling.PrintError(err)
		return models.Message{}, fmt.Errorf("error parsing player ID: %v", err)
	}
	start += playerNicknameSize

	//Read Parameters
	parametersStr := input[start:]

	params, err := parseParamsStr(parametersStr)
	if err != nil {
		errorHandeling.PrintError(err)
		return models.Message{}, fmt.Errorf("error parsing params: %v", err)
	}

	//convert values
	commandIDint, err := strconv.ParseUint(commandID, 10, 64)
	if err != nil {
		errorHandeling.PrintError(err)
		return models.Message{}, fmt.Errorf("error parsing command ID: %v", err)
	}

	return models.Message{
		Signature:      signature, // You can populate this with the actual signature values
		CommandID:      int(commandIDint),
		TimeStamp:      timeStamp,
		PlayerNickname: playerNickname,
		Parameters:     params,
	}, nil
}

/*
ConvertParamClientCreateGame

	checks if the parameters are correct and converts them to the correct type
*/
func ConvertParamClientCreateGame(params []constants.Params, names []string) (string, int, error) {
	gameName := ""
	maxPlayers := -1
	var err error

	if len(params) != len(names) {
		return gameName, maxPlayers, fmt.Errorf("invalid number of arguments")
	}

	for i := 0; i < len(params); i++ {
		current := params[i]
		if current.Name != names[i] {
			return gameName, maxPlayers, fmt.Errorf("invalid number of arguments")
		}

		switch current.Name {
		case "gameName":
			gameName = current.Value
		case "maxPlayers":
			maxPlayers, err = strconv.Atoi(current.Value)
			if err != nil {
				errorHandeling.PrintError(err)
				return gameName, maxPlayers, fmt.Errorf("invalid number of arguments")
			}
		default:
			return gameName, maxPlayers, fmt.Errorf("invalid number of arguments")
		}
	}

	if gameName == "" || maxPlayers == -1 {
		err := fmt.Errorf("values havent been converted")
		errorHandeling.PrintError(err)
		return gameName, maxPlayers, err
	}

	return gameName, maxPlayers, nil
}

func ConvertParamClientJoinGame(params []constants.Params, names []string) (string, error) {
	gameName := ""

	if len(params) != len(names) {
		return gameName, fmt.Errorf("invalid number of arguments")
	}

	for i := 0; i < len(params); i++ {
		current := params[i]
		if current.Name != names[i] {
			return gameName, fmt.Errorf("invalid number of arguments")
		}

		switch current.Name {
		case "gameName":
			gameName = current.Value
			if !models.IsValidName(gameName) {
				err := fmt.Errorf("invalid game name")
				errorHandeling.PrintError(err)
				return gameName, err
			}
		default:
			return gameName, fmt.Errorf("invalid number of arguments")
		}
	}

	return gameName, nil
}

func ConvertParamClientSelectedCubes(params []constants.Params, names []string) ([]int, error) {
	//nested function isValidCubeValueList
	isValidCubeValueList := func(cubeValue int) bool {

		for _, scoreCube := range constants.CGScoreCubeValues {
			if cubeValue == scoreCube.Value {
				return true
			}
		}

		return false
	}

	var cubeValueList []int

	if len(params) != len(names) {
		return cubeValueList, fmt.Errorf("invalid number of arguments")
	}

	for i := 0; i < len(params); i++ {
		current := params[i]
		if current.Name != names[i] {
			return cubeValueList, fmt.Errorf("invalid number of arguments")
		}

		switch current.Name {
		case "cubeValues":
			array, err := parseParamValueArray(current.Value)
			if err != nil {
				errorHandeling.PrintError(err)
				return cubeValueList, fmt.Errorf("invalid number of arguments")
			}
			for _, value := range array {
				intValue, err := strconv.Atoi(value)
				if err != nil {
					errorHandeling.PrintError(err)
					return cubeValueList, fmt.Errorf("invalid number of arguments")
				}

				if !isValidCubeValueList(intValue) {
					err := fmt.Errorf("not valid cube values")
					errorHandeling.PrintError(err)
					return cubeValueList, err
				}

				cubeValueList = append(cubeValueList, intValue)
			}
		default:
			return cubeValueList, fmt.Errorf("invalid number of arguments")
		}
	}

	return cubeValueList, nil
}

func parseParamValueArray(value string) ([]string, error) {
	elementName := "value"

	var valueArray []string

	if len(value) == 0 {
		return valueArray, nil
	}

	if value[0] != constants.CArrayBrackets.Opening[0] {
		return valueArray, fmt.Errorf("invalid valueArray format")
	}

	if value[len(value)-1] != constants.CArrayBrackets.Closing[0] {
		return valueArray, fmt.Errorf("invalid valueArray format")
	}

	value = value[1 : len(value)-1]

	valuesStr := strings.Split(value, constants.CParamsListElementDelimiter)

	for _, valueStr := range valuesStr {
		paramsArray, err := parseParamsStr(valueStr)
		if err != nil {
			errorHandeling.PrintError(err)
			return valueArray, fmt.Errorf("invalid valueArray format")
		}
		if len(paramsArray) != 1 {
			err := fmt.Errorf("invalid valueArray format")
			errorHandeling.PrintError(err)
			return valueArray, err
		}
		if paramsArray[0].Name != elementName {
			err := fmt.Errorf("invalid valueArray format")
			errorHandeling.PrintError(err)
			return valueArray, err
		}

		parsedValue := paramsArray[0].Value

		valueArray = append(valueArray, parsedValue)
	}

	return valueArray, nil
}

func parseParamsStr(paramsString string) ([]constants.Params, error) {
	var paramArray []constants.Params

	if len(paramsString) == 0 {
		return paramArray, nil
	}

	if paramsString[0] != constants.CParamsBrackets.Opening[0] {
		err := fmt.Errorf("invalid paramArray format")
		errorHandeling.PrintError(err)
		return paramArray, err
	}

	if paramsString[len(paramsString)-1] != constants.CParamsBrackets.Closing[0] {
		err := fmt.Errorf("invalid paramArray format")
		errorHandeling.PrintError(err)
		return paramArray, err
	}

	paramsString = paramsString[1 : len(paramsString)-1]

	paramsStr := strings.Split(paramsString, constants.CParamsDelimiter)

	for _, paramStr := range paramsStr {
		parameter, err := parseParam(paramStr)
		if err != nil {
			errorHandeling.PrintError(err)
			return paramArray, fmt.Errorf("invalid paramArray format")
		}
		paramArray = append(paramArray, parameter)
	}

	return paramArray, nil
}

func parseParam(str string) (constants.Params, error) {
	parameter := constants.Params{}

	if len(str) == 0 {
		return parameter, nil
	}

	paramKeyValue := strings.SplitN(str, constants.CParamsKeyValueDelimiter, 2)

	parameter.Name = paramKeyValue[0]
	if parameter.Name[0] != constants.CParamsWrapper[0] {
		return parameter, fmt.Errorf("invalid param format")
	}

	if parameter.Name[len(parameter.Name)-1] != constants.CParamsWrapper[0] {
		return parameter, fmt.Errorf("invalid param format")
	}
	parameter.Name = parameter.Name[1 : len(parameter.Name)-1]

	parameter.Value = paramKeyValue[1]
	if parameter.Value[0] != constants.CParamsWrapper[0] {
		return parameter, fmt.Errorf("invalid param format")
	}
	if parameter.Value[len(parameter.Value)-1] != constants.CParamsWrapper[0] {
		return parameter, fmt.Errorf("invalid param format")
	}
	parameter.Value = parameter.Value[1 : len(parameter.Value)-1]

	return parameter, nil
}

func parseMessagePlayerID(input string) (string, int, error) {
	playerNickname := ""
	playerNicknameSize := 0

	if len(input) == 0 {
		return playerNickname, playerNicknameSize, fmt.Errorf("invalid player ID format")
	}

	opening := constants.CParamsBrackets.Opening
	closing := constants.CParamsBrackets.Closing

	playerNickname, err := extractSubstring(input, opening, closing)
	if err != nil {
		errorHandeling.PrintError(err)
		return playerNickname, playerNicknameSize, fmt.Errorf("invalid player ID format")
	}

	playerNicknameSize = len(playerNickname) + len(opening) + len(closing)

	return playerNickname, playerNicknameSize, nil
}

// endregion

// region FUNCTIONS UTILS
func extractSubstring(input, opening, closing string) (string, error) {
	startIndex := strings.Index(input, opening)
	if startIndex == -1 {
		return "", fmt.Errorf("opening character not found")
	}
	endIndex := strings.Index(input[startIndex:], closing)
	if endIndex == -1 {
		return "", fmt.Errorf("closing character not found")
	}
	return input[startIndex+len(opening) : startIndex+endIndex], nil
}

func getArrayElementStr(name string, value string) string {
	returnStr := constants.CParamsWrapper + name + constants.CParamsWrapper + constants.CParamsKeyValueDelimiter + constants.CParamsWrapper + value + constants.CParamsWrapper
	return returnStr
}

// endregion

// region FUNCTIONS CONVERT TO NETWORK STRING

func ConvertListGameListToNetworkString(array []*models.Game) string {
	fieldOrder := []string{"gameName", "maxPlayers", "connectedPlayers"}
	return convertListToNetworkString(array, func(element interface{}) map[string]string {
		game := element.(*models.Game)
		return map[string]string{
			"gameName":         game.GetName(),
			"maxPlayers":       fmt.Sprintf("%d", game.GetMaxPlayers()),
			"connectedPlayers": fmt.Sprintf("%d", game.GetPlayersCount()),
		}
	}, fieldOrder)
}

func ConvertListPlayerListToNetworkString(array []*models.Player) string {
	fieldOrder := []string{"playerName", "isConnected"}
	return convertListToNetworkString(array, func(element interface{}) map[string]string {
		player := element.(*models.Player)
		return map[string]string{
			"playerName":  player.GetNickname(),
			"isConnected": convertBoolToNetworkString(player.IsConnected()),
		}
	}, fieldOrder)
}

func ConvertListGameDataToNetworkString(data models.GameData) string {
	fieldOrder := []string{"playerName", "isConnected", "score", "isTurn"}
	return convertListToNetworkString(data.PlayerGameDataArr, func(element interface{}) map[string]string {
		playerGameData := element.(models.PlayerGameData)
		return map[string]string{
			"playerName":  playerGameData.Player.GetNickname(),
			"isConnected": convertBoolToNetworkString(playerGameData.Player.IsConnected()),
			"score":       fmt.Sprintf("%d", playerGameData.Score),
			"isTurn":      convertBoolToNetworkString(playerGameData.Player == data.TurnPlayer),
		}
	}, fieldOrder)
}

func ConvertListCubeValuesToNetworkString(values []int) string {
	fieldOrder := []string{"value"}
	return convertListToNetworkString(values, func(element interface{}) map[string]string {
		value := element.(int)
		return map[string]string{
			"value": fmt.Sprintf("%d", value),
		}
	}, fieldOrder)
}

func ConvertMessageToNetworkString(message models.Message) (string, error) {
	networkString := ""

	messageData := models.MessageHeader{}
	signatureSize := int(unsafe.Sizeof(messageData.Signature))
	commandIDSize := int(unsafe.Sizeof(messageData.CommandID))
	timeStempSize := int(unsafe.Sizeof(messageData.TimeStamp))

	//Signature
	if len(message.Signature) != signatureSize {
		return networkString, fmt.Errorf("invalid signature")
	}
	networkString += message.Signature

	//Command ID
	commandIDStr := fmt.Sprintf("%02d", message.CommandID)
	if len(commandIDStr) != commandIDSize {
		return networkString, fmt.Errorf("invalid command ID")
	}
	networkString += commandIDStr

	//TimeStemp
	if len(message.TimeStamp) != timeStempSize {
		return networkString, fmt.Errorf("invalid time stamp")
	}
	networkString += message.TimeStamp

	//PlayerNickname
	networkString += convertPlayerNicknameToNetworkString(message.PlayerNickname)

	//Parameters
	paramsStr, err := convertParamsToNetworkString(message.Parameters)
	if err != nil {
		errorHandeling.PrintError(err)
		return networkString, err
	}
	networkString += paramsStr

	//End delimiter
	networkString += string(constants.CMessageEndDelimiter)

	return networkString, nil
}

func convertListToNetworkString(array interface{}, extractFields func(interface{}) map[string]string, fieldOrder []string) string {
	arrayValue := reflect.ValueOf(array)
	if arrayValue.Kind() != reflect.Slice {
		return ""
	}

	arrayStr := constants.CArrayBrackets.Opening

	for i := 0; i < arrayValue.Len(); i++ {
		element := arrayValue.Index(i).Interface()
		fields := extractFields(element)

		arrayStr += constants.CParamsBrackets.Opening
		for j, name := range fieldOrder {
			if j > 0 {
				arrayStr += constants.CParamsDelimiter
			}
			arrayStr += getArrayElementStr(name, fields[name])
		}
		arrayStr += constants.CParamsBrackets.Closing

		if i < arrayValue.Len()-1 {
			arrayStr += constants.CParamsListElementDelimiter
		}
	}

	arrayStr += constants.CArrayBrackets.Closing
	return arrayStr
}

//func ConvertListGameListToNetworkString(array []*models.Game) string {
//	nameGameName := "gameName"
//	nameMaxPlayers := "maxPlayers"
//	nameConnectedPlayers := "connectedPlayers"
//
//	arrayStr := ""
//
//	arrayStr += constants.CArrayBrackets.Opening
//
//	for i, game := range array {
//
//		valueGameName := game.GetName()
//		valueMaxPlayers := fmt.Sprintf("%d", game.GetMaxPlayers())
//		valueConnectedPlayers := fmt.Sprintf("%d", game.GetPlayersCount())
//
//		arrayStr += constants.CParamsBrackets.Opening
//		arrayStr += getArrayElementStr(nameGameName, valueGameName)
//		arrayStr += constants.CParamsDelimiter
//		arrayStr += getArrayElementStr(nameMaxPlayers, valueMaxPlayers)
//		arrayStr += constants.CParamsDelimiter
//		arrayStr += getArrayElementStr(nameConnectedPlayers, valueConnectedPlayers)
//		arrayStr += constants.CParamsBrackets.Closing
//
//		if i < len(array)-1 {
//			arrayStr += constants.CParamsListElementDelimiter
//		}
//	}
//
//	arrayStr += constants.CArrayBrackets.Closing
//
//	return arrayStr
//}
//
//func ConvertListPlayerListToNetworkString(array []*models.Player) string {
//	namePlayerName := "playerName"
//	nameIsConnected := "isConnected"
//
//	arrayStr := ""
//
//	arrayStr += constants.CArrayBrackets.Opening
//
//	for i, player := range array {
//
//		valuePlayerName := player.GetNickname()
//		//valueIsConnected := fmt.Sprintf("%t", player.IsConnected())
//		valueIsConnected := convertBoolToNetworkString(player.IsConnected())
//
//		arrayStr += constants.CParamsBrackets.Opening
//		arrayStr += getArrayElementStr(namePlayerName, valuePlayerName)
//		arrayStr += constants.CParamsDelimiter
//		arrayStr += getArrayElementStr(nameIsConnected, valueIsConnected)
//		arrayStr += constants.CParamsBrackets.Closing
//
//		if i < len(array)-1 {
//			arrayStr += constants.CParamsDelimiter
//		}
//	}
//
//	arrayStr += constants.CArrayBrackets.Closing
//
//	return arrayStr
//}
//
//func ConvertListGameDataToNetworkString(data models.GameData) string {
//
//	namePlayerName := "playerName"
//	nameIsConnected := "isConnected"
//	nameScore := "score"
//	nameIsTurn := "isTurn"
//
//	arrayStr := ""
//	arrayStr += constants.CArrayBrackets.Opening
//
//	array := data.PlayerGameDataArr
//	for i, playerGameData := range array {
//		valuePlayerName := playerGameData.Player.GetNickname()
//		valueIsConnected := convertBoolToNetworkString(playerGameData.Player.IsConnected())
//		valueScore := fmt.Sprintf("%d", playerGameData.Score)
//		valueIsTurn := convertBoolToNetworkString(playerGameData.Player == data.TurnPlayer)
//
//		arrayStr += constants.CParamsBrackets.Opening
//		arrayStr += getArrayElementStr(namePlayerName, valuePlayerName)
//		arrayStr += constants.CParamsDelimiter
//		arrayStr += getArrayElementStr(nameIsConnected, valueIsConnected)
//		arrayStr += constants.CParamsDelimiter
//		arrayStr += getArrayElementStr(nameScore, valueScore)
//		arrayStr += constants.CParamsDelimiter
//		arrayStr += getArrayElementStr(nameIsTurn, valueIsTurn)
//		arrayStr += constants.CParamsBrackets.Closing
//
//		if i < len(array)-1 {
//			arrayStr += constants.CParamsDelimiter
//		}
//	}
//
//	arrayStr += constants.CArrayBrackets.Closing
//	return arrayStr
//}
//
//func ConvertListCubeValuesToNetworkString(values []int) string {
//	nameValue := "value"
//
//	arrayStr := ""
//
//	arrayStr += constants.CArrayBrackets.Opening
//
//	for i, value := range values {
//
//		valueValue := fmt.Sprintf("%d", value)
//
//		arrayStr += constants.CParamsBrackets.Opening
//		arrayStr += getArrayElementStr(nameValue, valueValue)
//		arrayStr += constants.CParamsBrackets.Closing
//		if i < len(values)-1 {
//			arrayStr += constants.CParamsDelimiter
//		}
//	}
//
//	arrayStr += constants.CArrayBrackets.Closing
//
//	return arrayStr
//}

func convertBoolToNetworkString(value bool) string {
	//true to 1, false to 0
	if value {
		return "1"
	}
	return "0"
}

func convertPlayerNicknameToNetworkString(nickname string) string {
	networkStr := ""
	networkStr += constants.CParamsBrackets.Opening
	networkStr += nickname
	networkStr += constants.CParamsBrackets.Closing

	return networkStr
}

func convertParamsToNetworkString(params []constants.Params) (string, error) {
	length := len(params)
	if length == 0 {
		return constants.CParamsBrackets.Opening + constants.CParamsBrackets.Closing, nil
	}

	paramsStr := constants.CParamsBrackets.Opening
	for i, param := range params {
		paramStr := _convertParamElementToNetworkString(param.Name, param.Value)
		paramsStr += paramStr
		if i < length-1 {
			paramsStr += constants.CParamsDelimiter
		}
	}

	paramsStr += constants.CParamsBrackets.Closing

	return paramsStr, nil
}

func _convertParamElementToNetworkString(name string, value string) string {
	return constants.CParamsWrapper + name + constants.CParamsWrapper + constants.CParamsKeyValueDelimiter + constants.CParamsWrapper + value + constants.CParamsWrapper
}

// endregion
